File: rudolph/src/main/java/cn/wzbos/android/rudolph/router/ServiceRouter.java
package cn.wzbos.android.rudolph.router;


import android.util.Log;

import java.lang.reflect.Constructor;

import cn.wzbos.android.rudolph.IRouteService;
import cn.wzbos.android.rudolph.exception.ErrorCode;
import cn.wzbos.android.rudolph.exception.ErrorMessage;
import cn.wzbos.android.rudolph.exception.RudolphException;

public class ServiceRouter<R> extends Router<R> {

    ServiceRouter(RouteBuilder<?, ?> builder) {
        super(builder);
    }

    protected ServiceRouter(Builder<?, ? extends IRouteService> builder) {
        super(builder);
    }


    @Override
    @SuppressWarnings("unchecked")
    public R open() {
        if (super.intercept(null))
            return null;

        if (target == null) {
            if (callback != null)
                callback.onError(this, new RudolphException(ErrorCode.NOT_FOUND, ErrorMessage.NOT_FOUND_ERROR));
            return null;
        }

        try {
            Constructor<?> constructor = target.getConstructor();
            if (constructor.getParameterTypes().length > 0) {
                if (callback != null)
                    callback.onError(this, new RudolphException(ErrorCode.SERVICE_CREATE_FAILED, "创建服务失败,请提供一个无参数构造方法！"));
            }

            if (!constructor.isAccessible()) {
                Log.w("rudolph", "警告：构造方法 " + target.getName() + " 为私有!");
                constructor.setAccessible(true);
            }
            Object instance = constructor.newInstance();
            if (instance instanceof IRouteService) {
                IRouteService component = ((IRouteService) instance);
                component.init(this.bundle);
                if (callback != null)
                    callback.onSuccess(this);
                return (R) component;
            }
        } catch (Exception e) {
            if (callback != null) {
                callback.onError(this, new RudolphException(ErrorCode.SERVICE_CREATE_FAILED, "创建服务失败！", e));
            } else {
                Log.e("rudolph", "创建服务失败!");
            }
        }
        return null;
    }


    public static class Builder<T extends Builder<T, R>, R extends IRouteService> extends RouteBuilder<Builder<T, R>, ServiceRouter<R>> {

        public Builder(Class<?> target) {
            super(target);
        }

        public Builder(String path) {
            super(path);
        }

        @Override
        public ServiceRouter<R> build() {
            return new ServiceRouter<>(this);
        }
    }
}

File: rudolph/src/main/java/cn/wzbos/android/rudolph/router/ServiceRouter.kt
package cn.wzbos.android.rudolph.router

import cn.wzbos.android.rudolph.IRouteService
import cn.wzbos.android.rudolph.exception.ErrorCode
import cn.wzbos.android.rudolph.exception.ErrorMessage
import cn.wzbos.android.rudolph.exception.RudolphException
import cn.wzbos.android.rudolph.logger.RLog
import java.lang.reflect.Constructor

@Suppress("UNCHECKED_CAST")
open class ServiceRouter<R> : Router<R?> {
    protected constructor(builder: Builder<*, out IRouteService?>) : super(builder)
    constructor(builder: UriRouter.Builder<*>) : super(builder)

    override fun execute(): R? {
        if (super.intercept(null)) return null
        if (target == null) {
            callback?.onError(
                this,
                RudolphException(ErrorCode.NOT_FOUND, ErrorMessage.NOT_FOUND_ERROR)
            )
            return null
        }
        try {
            val constructor: Constructor<*> = target!!.getConstructor()
            if (constructor.parameterTypes.isNotEmpty()) {
                callback?.onError(
                    this,
                    RudolphException(ErrorCode.SERVICE_CREATE_FAILED, "创建服务失败,请提供一个无参数构造方法！")
                )
            }
            if (!constructor.isAccessible) {
                constructor.isAccessible = true
            }
            val instance: Any = constructor.newInstance()
            if (instance is IRouteService) {
                val component: IRouteService = instance
                component.init(extras)
                callback?.onSuccess(this)
                return component as R
            }
        } catch (e: Exception) {
            if (callback != null) {
                callback?.onError(
                    this,
                    RudolphException(ErrorCode.SERVICE_CREATE_FAILED, "创建服务失败！", e)
                )
            } else {
                RLog.e("ServiceRouter", "创建服务失败!")
            }
        }
        return null
    }

    open class Builder<T : Builder<T, R>?, R : IRouteService?>(path: String) :
        RouteBuilder<Builder<T, R>?, ServiceRouter<R>?>(path) {

        override fun build(): ServiceRouter<R> {
            return ServiceRouter(this)
        }
    }
}